### PR TITLE
Don't impose a strict size limit to uploads

### PR DESCRIPTION
**Pull request by sigurdm**
Fix of: https://github.com/dart-lang/pub/issues/4013

### CODE REVIEW

**Review by szakarias — APPROVED** [review]
LGTM

**Comment by sigurdm** [comment]
When testing uploads of too large packages, the error message was not descriptive. So I added a poor person's parsing of the error messages from GCS.

PTAL.

File: lib/src/http.dart [hunk 1]
```diff
@@ -211,6 +211,40 @@ void handleJsonError(http.BaseResponse response) {
   fail(log.red(error['message'] as String));
 }
 
+/// Handles an unsuccessful XML-formatted response from google cloud storage.
+///
+/// Assumes messages are of the form in
+/// https://cloud.google.com/storage/docs/xml-api/reference-status.
```
Comment (jonasfj) [inline lib/src/http.dart:217]:
```suggestion
/// https://cloud.google.com/storage/docs/xml-api/reference-status
```

File: test/lish/cloud_storage_upload_provides_an_error_test.dart [hunk 2]
```diff
@@ -22,14 +22,30 @@ void main() {
...
+    expect(
+      pub.stderr,
+      emits(
+        'Server message: Your proposed upload is larger than the maximum object size specified in your Policy Document.',
+      ),
+    );
+    expect(
+      pub.stderr,
+      emits('Server details: Content-length exceeds upper bound on range'),
+    );
```
Comment (jonasfj) [inline test/lish/cloud_storage_upload_provides_an_error_test.dart:47]: What does this look like in a terminal?

Are you sure we don't want to special case this?

**Review by jonasfj — APPROVED** [review]

File: test/lish/cloud_storage_upload_provides_an_error_test.dart (hunk 2 above)
Comment (sigurdm) [reply test/lish/cloud_storage_upload_provides_an_error_test.dart:47]: > What does this look like in a terminal?

image

> Are you sure we don't want to special case this?

Such that we recognize this specific message and give our own reply? That might work as well...

Comment (sigurdm) [reply test/lish/cloud_storage_upload_provides_an_error_test.dart:47]: I think special-casing for specific messages would be neat - but also risks swallowing specific details from the error message.

I'll add a TODO for now.

**Pull request merged**